bin/packages/lint-staged-typecheck.js

```diff
@@ -0,0 +1,34 @@
+/**
+ * External dependencies
+ */
+const _ = require( 'lodash' );
+const path = require( 'path' );
+const fs = require( 'fs' );
+const execa = require( 'execa' );
+
+/* eslint-disable no-console */
+
+const repoRoot = path.join( __dirname, '..', '..' );
+const tscPath = path.join( repoRoot, 'node_modules', '.bin', 'tsc' );
+
+// lint-staged passes full paths to staged changes
+const changedFiles = process.argv.slice( 2 );
+
+// Transform changed files to package directories containing tsconfig.json
+const changedPackages = _.uniq(
+	changedFiles.map( ( fullPath ) => {
+		const relativePath = path.relative( repoRoot, fullPath );
+		return path.join( ...relativePath.split( path.sep ).slice( 0, 2 ) );
+	} )
+).filter( ( packageRoot ) =>
+	fs.existsSync( path.join( packageRoot, 'tsconfig.json' ) )
+);
+
+try {
+	execa.sync( tscPath, [ '--build', ...changedPackages ] );
```

Comment:
If we keep this script lint-staged, this could invoke `build:package-types` script with `-- ...changedPackages`.

I'd prefer to remove this entirely as noted: https://github.com/WordPress/gutenberg/pull/18942/files#r396317889